lib/tenstorrent/bh_arc/pvt.c

```diff
@@ -19,594 +18,114 @@
+		     {SENSOR_CHAN_PVT_TT_BH_VM, 3}, {SENSOR_CHAN_PVT_TT_BH_VM, 4},
+		     {SENSOR_CHAN_PVT_TT_BH_VM, 5}, {SENSOR_CHAN_PVT_TT_BH_VM, 6},
+		     {SENSOR_CHAN_PVT_TT_BH_VM, 7});
 
-/* can not readback supply check in auto mode, use manual read instead */
-static ReadStatus ReadVM(uint32_t id, uint16_t *data)
-{
-	/* ignore ip_done in auto_mode */
-	IpDataRegU ip_data;
-
-	ip_data.val = ReadReg(GET_VM_REG_ADDR(id, SDIF_RDATA));
-
-	if (ip_data.f.ip_fault) {
-		return SampleFault;
-	}
-	if (ip_data.f.ip_type != ValidData) {
-		return IncorrectSampleType;
-	}
-	*data = ip_data.f.ip_dat;
-	return ReadOk;
-}
+SENSOR_DT_READ_IODEV(ts_iodev, DT_NODELABEL(pvt), {SENSOR_CHAN_PVT_TT_BH_TS, 0},
+		     {SENSOR_CHAN_PVT_TT_BH_TS, 1}, {SENSOR_CHAN_PVT_TT_BH_TS, 2},
+		     {SENSOR_CHAN_PVT_TT_BH_TS, 3}, {SENSOR_CHAN_PVT_TT_BH_TS, 4},
+		     {SENSOR_CHAN_PVT_TT_BH_TS, 5}, {SENSOR_CHAN_PVT_TT_BH_TS, 6},
+		     {SENSOR_CHAN_PVT_TT_BH_TS, 7});
 
-static ReadStatus ReadPD(uint32_t id, uint16_t *data)
-{
-	return ReadPvtAutoMode(PD, id, data, PVT_CNTL_PD_00_SDIF_DONE_REG_ADDR,
-			       PVT_CNTL_PD_00_SDIF_DATA_REG_ADDR);
-}
+SENSOR_DT_READ_IODEV(pd_iodev, DT_NODELABEL(pvt), {SENSOR_CHAN_PVT_TT_BH_PD, 0},
+		     {SENSOR_CHAN_PVT_TT_BH_PD, 1}, {SENSOR_CHAN_PVT_TT_BH_PD, 2},
+		     {SENSOR_CHAN_PVT_TT_BH_PD, 3}, {SENSOR_CHAN_PVT_TT_BH_PD, 4},
+		     {SENSOR_CHAN_PVT_TT_BH_PD, 5}, {SENSOR_CHAN_PVT_TT_BH_PD, 6},
+		     {SENSOR_CHAN_PVT_TT_BH_PD, 7}, {SENSOR_CHAN_PVT_TT_BH_PD, 8},
+		     {SENSOR_CHAN_PVT_TT_BH_PD, 9}, {SENSOR_CHAN_PVT_TT_BH_PD, 10},
+		     {SENSOR_CHAN_PVT_TT_BH_PD, 11}, {SENSOR_CHAN_PVT_TT_BH_PD, 12},
+		     {SENSOR_CHAN_PVT_TT_BH_PD, 13}, {SENSOR_CHAN_PVT_TT_BH_PD, 14},
+		     {SENSOR_CHAN_PVT_TT_BH_PD, 15});
 
-float GetAvgChipTemp(void)
-{
-	float ts_sum = 0;
-
-	for (uint32_t i = 0; i < NUM_TS; i++) {
-		uint16_t tmon_data;
-		ReadStatus pd_read_status;
-
-		pd_read_status = ReadTS(i, &tmon_data);
-		if (pd_read_status != ReadOk) {
-			/* Something went wrong! Return -FLT_MAX to indicate an error */
-			return -FLT_MAX;
-		}
-		ts_sum += DoutToTemp(tmon_data);
-	}
-	return ts_sum / NUM_TS;
-}
+RTIO_DEFINE(pvt_ctx, 16, 16);
 
-static void SelectDelayChainAndStartPDConv(uint32_t delay_chain)
-{
-	if (delay_chain != selected_pd_delay_chain) {
-		PDIpCfg0U ip_cfg0;
-
-		ip_cfg0.val = 0;
-		ip_cfg0.f.run_mode = 0; /* MODE_PD_CNV */
-		ip_cfg0.f.oscillator_enable = ALL_AGING_OSC;
-		ip_cfg0.f.oscillator_select = delay_chain;
-		ip_cfg0.f.counter_gate = 0x3; /* W = 255 */
-		SdifWrite(PVT_CNTL_PD_CMN_SDIF_STATUS_REG_ADDR, PVT_CNTL_PD_CMN_SDIF_REG_ADDR,
-			  IP_CFG0_ADDR, ip_cfg0.val);
-		SdifWrite(PVT_CNTL_PD_CMN_SDIF_STATUS_REG_ADDR, PVT_CNTL_PD_CMN_SDIF_REG_ADDR,
-			  IP_CNTL_ADDR, 0x108); /* ip_run_cont */
-
-		/* wait until delay chain takes effect */
-		WaitUs(250);
-		selected_pd_delay_chain = delay_chain;
-	}
-}
+/*
+ * Used for storing raw temperature sensor data when reading.
+ */
+uint8_t buf[sizeof(struct sensor_value) * 8];
```

Comment:
Is there a way to get the number of channels from `ts_iodev` programmatically?

Reply:
Ah yes there is, I missed that